File: oa_questions/arrays/4MediumOfTwoSortedArrays.py
# hard
# 4MediumOfTwoSortedArrays
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        copy1, copy2 = nums1, nums2
        total = len(nums1) + len(nums2)
        half = total // 2

        if len(copy2) < len(copy1):
             copy1, copy2 = copy2, copy1
        
        l, r = 0, len(copy1) - 1
        while True:
            i = (l + r) // 2 # index of copy1
            j = half - i - 2 # index of copy2

            oneLeft  = copy1[i] if i >= 0 else float("-infinity")
            oneRight = copy1[i+1] if (i+1) < len(copy1) else float("infinity")
            twoLeft  = copy2[j] if j >= 0 else float("-infinity")
            twoRight = copy2[j+1] if (j+1) < len(copy2) else float("infinity")

            # condition of partition correct
            if oneLeft <= twoRight and twoLeft <= oneRight:
                # odd
                if total % 2:
                    return min(oneRight, twoRight)
                #even
                return (max(oneLeft, twoLeft) + min(oneRight, twoRight)) / 2
            elif oneLeft > twoRight:
                r = i - 1
            else:
                l = i + 1

File: oa_questions/arrays/4MediumOfTwoSortedArrays.java
// hard
// 4MediumOfTwoSortedArrays
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        double median = 0;
        int medianIndex = 0;
        if (nums1.length == 0 && nums2.length > 0) {
            median = nums2[nums2.length/2];
        } else if (nums2.length == 0 && nums1.length > 0) {
            median = nums1[nums1.length/2];
        } else if (nums1.length == 0 && nums2.length == 0) {
            return 0;
        }
        
        int totalLength = nums1.length + nums2.length;
        medianIndex = (int) Math.ceil((double)(totalLength + 1)/2);
        int[] ans = new int[medianIndex];
        int iIndex = 0, jIndex = 0;
        for (int i = 0; i < medianIndex; i++) {
            int next = 0;
            if (iIndex < nums1.length && jIndex < nums2.length) {
                next = (nums1[iIndex] < nums2[jIndex])
                ? nums1[iIndex++]
                : nums2[jIndex++];
            } else if (jIndex < nums2.length) {
                next = nums2[jIndex++];
            } else { // One of them has to still be available.
                next = nums1[iIndex++];
            }
            ans[i] = next;
        }
        if (totalLength % 2 == 1) {
            median = ans[medianIndex-1];
        } else{
            median = (double) (ans[medianIndex-2] + ans[medianIndex-1])/2;
        }
    return median;
    }
}